Fix globbing bugs for gitignore/semgrepignore

mjambon opened the pull request:
The following globbing bugs were fixed, making semgrepignore behavior in osemgrep more correct:

* `a/*` should not match `a/`
* `*` should not match `.a`
* `a//b` should match `a/b` and vice-versa

Improvements made to the test framework along the way:

* Print the path(s) to the captured output when the expected output is missing (allows reviewing it before approval)
* Add a function `mask_temp_paths` to conveniently hide random file paths in test output for comparison against past runs

Edit: target selection over the semgrep repo has become 2x slower (now 3x slower than pysemgrep)! I'm looking into it. It's very interesting.

Performance update: I solved it by adding the option `--exclude-standard` to `git ls-files --others` that honors gitignore, just like pysemgrep. The downside is that we can't de-ignore git-ignored files using semgrepignore. I think it's fine since deignoring files in semgrepignore is a new feature anyway. Here's the kind of deignoring we can do in semgrepignore:
```
tests/*
!tests/*.py
```
but we can't have the first pattern in a `.gitignore` and the second pattern (deignoring) in a `.semgrepignore`.

Commit:
Complete the basic 'osemgrep ls' tests

Commit:
Add function to mask temporary file paths in test output

Commit:
Print path to captured output when expected output is missing.
Also, renamed "unchecked output" to "log" and other minor interface
changes.

Commit:
Update test snapshot

Commit:
Fix incorrect globbing where pattern 'a/*' was matching 'a/'
This switches the implementation from ocaml-re to PCRE because the
former doesn't support lookahead assertions.

Commit:
Log PCRE patterns in globbing tests

Commit:
Add a script to compare targets selected by pysemgrep and osemgrep

Commit:
Don't match leading periods in path segments with '*' or '?'

Commit:
Update changelog

Commit:
Update test snapshots

Commit:
Change the output format of --x-ls for enhanced readability

Commit:
Pass the '--exclude-standard' option to 'git ls-files --others'. This
results in much fewer targets to filter through but prevents
de-exclusions of git-ignored files using semgrepignore.

Commit:
Fix target selection tests

Commit:
Display timeout value in timeout error message

Commit:
Fix git command used to list tracked files (don't pass '--exclude-standard')

ajbt200128 commented:
> but we can't have the first pattern in a .gitignore and the second pattern (deignoring) in a .semgrepignore.

Honestly this feels preferable. If you're scanning something that's been gitignored (i.e. dynamic dependencies, build output files etc.) then you're probably doing something weird, and should be doing it another way. 99% of users I imagine will run semgrep on commit, or in CI which really should not depend on build artifacts/non committed dynamic stuff

ajbt200128 commented on cli/src/semgrep/run_scan.py:
```diff
@@ -312,9 +312,9 @@ def run_rules(
     targets = target_manager.get_files_for_language(None, product)
     for path in sorted(targets.kept):
-        print(f"+ {path}")
+        print(f"SEL {path}")
```
nit: SEL + IGN feels less clear to me  in meaning as a developer than + or -. Maybe we should just say "selected" and "ignored", or keep the + and -

mjambon replied:
Sure. `+` and `-` make it hard to read the diffs that add their own layer of `+` and `-`.

ajbt200128 commented on libs/alcotest_ext/core/Store.ml:
```diff
@@ -195,25 +195,26 @@ let names_of_checked_output (output : T.output_kind) : string list =
   | Separate_stdout_stderr -> [ stdout_filename; stderr_filename ]
 
-let has_unchecked_output (output : T.output_kind) : bool =
+let describe_unchecked_output (output : T.output_kind) : string option =
```
nice

ajbt200128 commented on libs/alcotest_ext/tests/Failing_test.ml:
```diff
@@ -24,6 +24,8 @@ consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
 non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
 |});
+    t "missing snapshot" ~checked_output:Stdout (fun () ->
+        print_endline "hello");
```
nit: why do we print "hello" here?

mjambon replied:
I guess it felt better to print something so as to resemble an ordinary use case, but you're right it's not necessary. I see how it can be distracting. I'll remove it.

ajbt200128 commented on libs/gitignore/Unit_gitignore.ml:
```diff
@@ -4,38 +4,39 @@
 
-(* ?? *)
-let test_list (files : t list) () =
+let concat_lines lines = String.concat "\n" lines ^ "\n"
```
nit: I think there's Common.unlines or Common2.unlines

ajbt200128 commented on libs/gitignore/Unit_gitignore.ml:
```diff
@@ -44,87 +45,77 @@ let test_filter (files : F.t list) selection () =
-               | Ignored -> printf "[OK] %s: ignored\n" (Ppath.to_string path));
-      assert (not !error))
+             match status with
+             | Not_ignored -> printf "SEL %s\n" (Ppath.to_string path)
```
Same nit here re: SEL + IGN

ajbt200128 commented on src/osemgrep/cli_scan/Ls_subcommand.ml:
```diff
@@ -10,14 +10,13 @@ let run ~target_roots ~targeting_conf:conf () =
   |> List.iter (fun (x : Fppath.t) ->
-         Out.put (spf "+ %s" (Fpath.to_string x.fpath)));
+         Out.put (spf "SEL %s" (Fpath.to_string x.fpath)));
   skipped
```
Same comment re: SEL + IGN

ajbt200128 commented on src/osemgrep/tests/Test_target_selection.ml:
```diff
@@ -2,45 +2,58 @@
+  in
+  Alcotest.(check int) "exit code" 0 (Exit_code.to_int exit_code)
+
+let concat_lines lines = String.concat "\n" lines ^ "\n"
```
Same nit re: Common.unlines

mjambon replied:
I'm going to keep this because it saves a documentation lookup, `String.concat` being commonly used and part of the OCaml standard library.

ajbt200128 reviewed (approved):
Probs wanna wait for @aryx since I don't understand the lexer stuff

Commit:
Revise --x-ls output format

Commit:
Report uncaught Python exceptions with a stack trace rather than
just the exception contents. This allows finding out where the error
occurred.

Sample output from `semgrep --test` I'm getting after accidentally
deleting a line:

```
--------------------------------------------------------------------------------
The following config files produced errors:
	python/flask/security/xss/audit/template-unquoted-attribute-var.yaml: Traceback (most recent call last):
  File "/home/martin/semgrep/cli/src/semgrep/test.py", line 304, in invoke_semgrep_multi
    output = semgrep.run_scan.run_scan_and_return_json(config, targets, **kwargs)
  File "/home/martin/semgrep/cli/src/semgrep/run_scan.py", line 771, in run_scan_and_return_json
    ) = run_scan(
  File "/home/martin/semgrep/cli/src/semgrep/run_scan.py", line 573, in run_scan
    ) = run_rules(
  File "/home/martin/semgrep/cli/src/semgrep/run_scan.py", line 217, in run_rules
    ) = core_runner.invoke_semgrep_core(
  File "/home/martin/semgrep/cli/src/semgrep/core_runner.py", line 1042, in invoke_semgrep_core
    ) = self._run_rules_direct_to_semgrep_core(
  File "/home/martin/semgrep/cli/src/semgrep/core_runner.py", line 1016, in _run_rules_direct_to_semgrep_core
    raise e
  File "/home/martin/semgrep/cli/src/semgrep/core_runner.py", line 983, in _run_rules_direct_to_semgrep_core
    return self._run_rules_direct_to_semgrep_core_helper(
  File "/home/martin/semgrep/cli/src/semgrep/core_runner.py", line 923, in _run_rules_direct_to_semgrep_core_helper
    outputs = core_matches_to_rule_matches(rules, core_output)
  File "/home/martin/semgrep/cli/src/semgrep/core_output.py", line 200, in core_matches_to_rule_matches
    rule_match = convert_to_rule_match(match)
  File "/home/martin/semgrep/cli/src/semgrep/core_output.py", line 174, in convert_to_rule_match
    isinstance(rule.product.value, out.Secrets),
NameError: free variable 'rule' referenced before assignment in enclosing scope

Not sending pseudonymous metrics since metrics are configured to OFF and registry usage is False
```

Commit:
Fix semgrep-core command emitted in a second pass to applying an autofix
in 'semgrep --test'. This causes the path-based target filtering found
in rules ('paths' field) to be ignored consistently.
This problem was revealed by recent fixes in the Glob module, which are
possibly incorrect but don't invalidate the present fix.

Commit:
Print debugging info to stderr rather than stdout so as to not interfere
with JSON output

Commit:
Fix bug in Glob library causing '**/a' to not match '/x/a'

Commit:
Simplify test
